### ISSUE TITLE
Finding the Words to Say: Hidden State Visualizations for Language Models

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11451]
Finding the Words to Say: Hidden State Visualizations for Language Models<br>
<br>
https://ift.tt/b7rwYKD<br>
<br>
Jay Alammar<br>
<br>

<div><div><div><div>
<div>
<p>By visualizing the hidden state between a model's layers, we can get some clues as to the model's "thought process".</p>
<div>
<div>
<figure><img src="https://jalammar.github.io/images/explaining/rankings-gpt2xl.png"><figcaption><strong>Figure: Finding the words to say</strong><br>
After a language model generates a sentence, we can visualize a view of how the model came by each word (column). Each row is a model layer. The value and color indicate the ranking of the output token at that layer. The darker the color, the higher the ranking. Layer 0 is at the top. Layer 47 is at the bottom.<br><strong>Model:</strong>GPT2-XL<br></figcaption></figure>
</div>
</div>
<p>Part 2: Continuing the pursuit of making Transformer language models more transparent, this article showcases a collection of visualizations to uncover mechanics of language generation inside a pre-trained language model. These visualizations are all created using <a href="https://www.eccox.io">Ecco</a>, the open-source package we're releasing
</p>
<p>In the first part of this series, <a href="https://jalammar.github.io/explaining-transformers/">Interfaces for Explaining Transformer Language Models</a>, we showcased interactive interfaces for input saliency and neuron activations. In this article, we will focus on the hidden state as it evolves from model layer to the next. By looking at the hidden states produced by every transformer decoder block, we aim to gleam information about how a language model arrived at a specific output token. This method is explored by Voita et al.<cite></cite>. Nostalgebraist <cite></cite>
presents compelling visual treatments showcasing the evolution of token rankings, logit scores, and softmax
probabilities for the evolving hidden state through the various layers of the model.
</p>
<h2>Recap: Transformer Hidden States</h2>
<p>The following figure recaps how a transformer language model works. How the layers result in a final hidden state. And how that final state is then projected to the output vocabulary which results in a score assigned to each token in
the model's vocabulary. We can see here the top scoring tokens when DistilGPT2 is fed the input sequence " 1, 1,
":</p>
<figure><img src="https://jalammar.github.io/images/explaining/transformer-language-model-steps.png"><figcaption><strong>Figure: Recap of transformer language models.</strong><br>
This figure shows how the model arrives at the top five output token candidates and their probability scores. This shows us that at the final layer, the
model is 59% sure the next token is ' 1', and that would be chosen as the output token by greedy decoding.
Other probable outputs include ' 2' with 18% probability (maybe we are counting) and ' 0' with 5%
probability (maybe we are counting down).
</figcaption></figure><p>
Ecco provides a view of the model's top scoring tokens and their probability scores.
</p>
<figure><pre><code><span># Generate one token to complete this input string
</span><span>output</span> <span>=</span> <span>lm</span><span>.</span><span>generate</span><span>(</span><span>" 1, 1, 1,"</span><span>,</span> <span>generate</span><span>=</span><span>1</span><span>)</span>

<span># Visualize
</span><span>output</span><span>.</span><span>layer_predictions</span><span>(</span><span>position</span><span>=</span><span>6</span><span>,</span> <span>layer</span><span>=</span><span>5</span><span>)</span></code></pre></figure><p>
Which would show the following breakdown of candidate output tokens and their probability scores:
</p>
<figure><img src="https://jalammar.github.io/images/explaining/prediction_scores.PNG"><figcaption><strong>Figure: Ten tokens with highest probabilities at the final layer of the model.</strong><br></figcaption></figure><h2>Scores after each layer</h2>
<p>Applying the same projection to internal hidden states of the model gives us a view of how the model's conviction
for the output scoring developed over the processing of the inputs. This projection of internal hidden states
gives us a sense of which layer contributed the most to elevating the scores (and hence ranking) of a certain
potential output token.</p>
<figure><img src="https://jalammar.github.io/images/explaining/predictions.PNG"><figcaption><strong>Figure: projecting inner hidden states to the model's vocabulary reveals cues of processing between layers.</strong><br></figcaption></figure><p>Viewing the evolution of the hidden states means that instead of looking only at the candidates output tokens from
projecting the final model state, we can look at the top scoring tokens after projecting the hidden state
resulting from each of the model's six layers.</p>
<p>
This visualization is created using the same method above with omitting the 'layer' argument (which we set to the final layer in the previous example, layer #5):
</p>
<figure><pre><code><span># Visualize the top scoring tokens after each layer
</span><span>output</span><span>.</span><span>layer_predictions</span><span>(</span><span>position</span><span>=</span><span>6</span><span>)</span></code></pre></figure><p>
Resulting in: </p>
<figure><img src="https://jalammar.github.io/images/explaining/predictions%20all%20layers.PNG"><figcaption><strong>Figure: Top scoring tokens after each of the model's six layers.</strong>
<br>
Each row shows the top ten predicted tokens obtained by projecting each hidden state to the output
vocabulary. The probability scores are shown in pink (obtained by passing logit scores through softmax). We
can see that <strong>Layer 0</strong> has no digits in its top ten predictions. <strong>Layer 1</strong>
gives the token ' 1' a 0.03%, probability which, while low, still ranks the token as the seventh highest
ranking token. Subsequent layers keep elevating the probability and ranking of ' 1', until <strong>the final
layer</strong> injects a bit more caution by reducing the probability from 100% to ~60%, still retaining the
token as the highest ranked in the model's output.<br><strong>Note:</strong> This figure is incorrect in showing 0 probability assigned to some tokens due to rounding. The current version of Ecco fixes this by showing '&lt;0.01%'.
</figcaption></figure><p>
You can experiment with these visualizations and experiment with them on your own input sentences at the following colab link:
</p>
<p><a href="https://colab.research.google.com/github/jalammar/ecco/blob/main/notebooks/Ecco_Output_Token_Scores.ipynb"><img src="https://jalammar.github.io/images/explaining/colab-badge.svg"></a>
</p>
<h3>Evolution of the selected token</h3>
<figure><img src="https://jalammar.github.io/images/explaining/logit_ranking_1.png"><figcaption><strong>The ranking of the token ' 1' after each layer</strong><br><strong>Layer 0</strong> elevated the token ' 1' to be the 31st highest scored token in the hidden state it
produced. <strong>Layers 1 and 2</strong> kept increasing the ranking (to 7 then 5 respectively). All the
<strong>following layers</strong> were sure this is the best token and gave it the top ranking spot.
</figcaption></figure><p>Another visual perspective on the evolving hidden states is to re-examine the hidden states after selecting an output
token to see how the hidden state after each layer ranked that token. This is one of the many perspectives
explored by Nostalgebraist <cite></cite>
and the one we think is a great first approach. In the figure on the side, we can see the ranking (out of
+50,0000 tokens in the model's vocabulary) of the token ' 1' where each row
indicates a layer's output.
</p>
<p>The same visualization can then be plotted for an entire generated sequence, where each column indicates a
generation step (and its output token), and each row the ranking of
the output token at each layer:
</p>
<figure><img src="https://jalammar.github.io/images/explaining/sequence_111_rankings.PNG"><figcaption><strong>Evolution of the rankings of the output sequence ' 1 , 1'</strong><br>
We can see that <strong>Layer 3</strong> is the point at which the model started to be certain of
the digit ' 1' as the output. <br><strong>When the output is to be a comma</strong>, Layer 0 usually ranks
the comma as 5. <br><strong>When the output is to be a ' 1'</strong>, Layer 0 is less certain, but still ranks the ' 1' token at
31 or 32.
Notice that every output token is ranked #1 after Layer 5. That is the definition of <strong>greedy
sampling</strong> -- the reason we selected this token is because it was ranked first.
</figcaption></figure><p>Let us demonstrate this visualization by presenting the following input to GPT2-Large:</p>
<figure></figure><p></p>
<p>Visualizaing the evolution of the hidden states sheds light on how various layers contribute to generating this sequence as we can see in the following figure:</p>
<figure><a href="https://jalammar.github.io/images/explaining/ranking-eu-gpt2.png" target="_blank"><img src="https://jalammar.github.io/images/explaining/ranking-eu-gpt2-thumb.png"></a><br><br><figcaption><strong>Figure: Hidden state evolution of an output sequence</strong> <br>
Click to open image in full resolution. The figure reveals:
<ul>
<li>Columns of solid pink corresponding to newlines and periods. Starting from Layer #0 and onwards, the model is certain early on of these tokens, indicating Layer #0's awareness of certain syntactic properties (and that later layers raise no objections).</li>
<li>
Columns where country names are predicted are very bright at the top and it's up to the last five layers to really come up with the appropriate token.
</li>
<li>
Columns tracking the incrementing number tend to be resolved at layer #9.
</li>
<li>
The model erroneously lists Chile in the list, not a EU country. But notice that the ranking of that token is 43 -- indicating the error is better attributed to our token sampling method rather than to the model itself. In the case of all other countries they were correct and in the top 3.
</li>
<li>
Aside from Chile, the rest of the countries are correct, but also follow the alphabetical order followed in the input sequence.
</li>
</ul></figcaption></figure><p><a href="https://colab.research.google.com/github/jalammar/ecco/blob/main/notebooks/Ecco_Evolution_of_Selected_Token.ipynb"><img src="https://jalammar.github.io/images/explaining/colab-badge.svg"></a></p>
<h3>Rankings of Other Tokens</h3>
<figure><img src="https://jalammar.github.io/images/explaining/watch_keys_cabinet.png"><figcaption><strong>Figure: Rankings of which token should go in the blank</strong> <br>
While the final output succeeds in assigning the correct number, the first five layers surprisingly fail at
identifying the correct number (by giving " is" a higher ranking than " are", which is the correct answer).
Examining attention or inner-layer saliency could reveal clues as to the reason.
</figcaption></figure><p>We are not limited to watching the evolution of only one (the selected) token for a specific position. There are
cases where we want to compare the rankings of multiple tokens <i>in the same position</i> regardless if the model selected them or not. </p>
<p>One such case is the number prediction task described by Linzen et al.<cite></cite>
which arises from the English language phenomenon of subject-verb agreement. In that task, we want to analyze the
model's capacity to encode <i>syntactic number</i> (whether the subject we're addressing is singular or plural)
and <i>syntactic subjecthood</i> (which subject in the sentence we're addressing).
</p>
<p>Put simply, fill-in the blank. The only acceptable answers are 1) <strong>is</strong> 2) <strong>are</strong>:
</p>
<p>The key<strong>s</strong> to the cabinet ______ </p>
<p>To answer correctly, one has to first determine whether we're describing the keys (possible subject #1) or the
cabinet (possible subject #2). Having decided it is the keys, the second determination would be whether it is
singular or plural.</p>
<figure><img src="https://jalammar.github.io/images/explaining/watch_key_cabinets.png"><figcaption>
The model is able to assign a higher ranking to <strong>is</strong>, which is the correct token. Every layer
in the model managed to rank " is" higher than " are". The ranking of " are" remains high, however, as far as
rankings are concerned (the delta in probability scores might indicate otherwise, however).
</figcaption></figure><p>Contrast your answer for the first question with the following variation:</p>
<p>The key to the cabinet<strong>s</strong> ______ </p>
<p>The figures in this section visualize the hidden-state evolution of the tokens " is" and " are". The numbers
in the cells are their ranking in the position of the blank (Both columns address the same position in the
sequence, they're not subsequent positions as was the case in the previous visualization).</p>
<p>The first figure (showing the rankings for the sequence "The keys to the cabinet") raises the question of why do five layers fail the task and only the final layer sets the record
straight. This is likely a similar effect to that observed in BERT of the final layer being the most
task-specific<cite></cite>. It is also worth investigating whether that capability of succeeding at the task is predominantly localized in
Layer 5, or if the Layer is only the final expression in a circuit<cite></cite>
spanning multiple layers which is especially sensitive to subject-verb agreement.
</p>
<h3>Probing for bias</h3>
<p>This method can shed light on questions of bias and where they might emerge in a model. The following figures, for example, probe for the model's gender expectation associated with different professions:</p>
<figure><img src="https://jalammar.github.io/images/explaining/doctor.png"><img src="https://jalammar.github.io/images/explaining/nurse.png"><figcaption><strong>Figure: Probing bias in the model's association of gender with professions - Doctor and nurse</strong><br>
The first five layers all rank " man" higher than " woman" for both professions. For the nursing profession, the final layer decisively elevates " woman" to a higher ranking than " man".
</figcaption></figure><p>More systemaic and nuanced examination of bias in contextualized word embeddings (another term for the vectors we've been referring to as "hidden states") can be found in <cite></cite>.</p>
<p><a href="https://colab.research.google.com/github/jalammar/ecco/blob/main/notebooks/Ecco_Comparing_Token_Rankings.ipynb"><img src="https://jalammar.github.io/images/explaining/colab-badge.svg"></a></p>
<h2>Your turn!</h2>
<p>
You can proceed to do your own experiments using Ecco and the three notebooks in this article:
</p>
<p>
You can report issues you run into at the Ecco's Github page. Feel free to share any interesting findings at the Ecco <a href="https://github.com/jalammar/ecco/discussions">Discussion</a> board. I invite you again to read <a href="https://www.lesswrong.com/posts/AcKRB8wDpdaN6v6ru/interpreting-gpt-the-logit-lens">Interpreting GPT the Logit Lens</a> and see the various ways the author examines such a visualization. I leave you with a small gallery of examples showcasing the responses of different models to different input prompts.
</p>
<h2>Gallery</h2>
<figure><figcaption><strong>Input:</strong> "Heathrow airport is located in the city of"<br><strong>Model:</strong> DistilGPT2
</figcaption><a href="https://jalammar.github.io/hidden-states/london_rankings.png"><img src="https://jalammar.github.io/images/explaining/london_rankings.png"></a><br><hr>
<figcaption><strong>Input:</strong> "Some of the most glorious historical attractions in Spain date from the period of Muslim rule, including The Mezquita, built as the Great Mosque of Cordoba and the Medina Azahara, also in Cordoba and now in ruins but still visitable as such and built as the Madinat al-Zahra, the Palace of al-Andalus; and the Alhambra in Granada, a splendid, intact palace. There are also two synagogues still standing that were built during the era of Muslim Spain: Santa Maria la Blanca in Toledo and the Synagogue of Cordoba, in the Old City. Reconquista and Imperial era"<br><strong>Model:</strong> DistilGPT2
</figcaption><a href="https://jalammar.github.io/images/explaining/ranking-cordoba.png"> <img src="https://jalammar.github.io/images/explaining/ranking-cordoba.png"></a><br><br><figcaption><strong>Model:</strong> GPT2-Large</figcaption><a href="https://jalammar.github.io/images/explaining/cordoba-gpt2.png"><img src="https://jalammar.github.io/images/explaining/cordoba-gpt2.png"></a><br><br><figcaption><strong>Model:</strong> GPT2-XL</figcaption><a href="https://jalammar.github.io/images/explaining/cordoba-gpt2xl.png"><img src="https://jalammar.github.io/images/explaining/cordoba-gpt2xl.png"></a><br><hr>
<figcaption><strong>Input:</strong> "The countires of the European Union are:\n1. Austria\n2. Belgium\n3.
Bulgaria\n4." <br><strong>Model:</strong> DistilGPT2
</figcaption><a href="https://jalammar.github.io/images/explaining/ranking-eu.png"><img src="https://jalammar.github.io/images/explaining/ranking-eu.png"></a><br><figcaption><strong>Model:</strong> GPT2-Large</figcaption><a href="https://jalammar.github.io/images/explaining/ranking-eu-gpt2.png"><img src="https://jalammar.github.io/images/explaining/ranking-eu-gpt2.png"></a><br><br><figcaption><strong>Model:</strong> GPT2-XL</figcaption><a href="https://jalammar.github.io/images/explaining/ranking-eu-gpt2xl.png"><img src="https://jalammar.github.io/images/explaining/ranking-eu-gpt2xl.png"></a>
</figure><h2>Acknowledgements</h2>
<p>This article was vastly improved thanks to feedback on earlier drafts provided by
Abdullah Almaatouq,
Anfal Alatawi,
Fahd Alhazmi,
Hadeel Al-Negheimish,
Isabelle Augenstein,
Jasmijn Bastings,
Najwa Alghamdi,
Pepa Atanasova, and
Sebastian Gehrmann.
</p>
<h2>References</h2>
<h2>Citation</h2>
<div>
<p>
If you found this work helpful for your research, please cite it as following:
</p>
<div>
<pre><code>Alammar, J. (2021). Finding the Words to Say: Hidden State Visualizations for Language Models [Blog post]. Retrieved from https://jalammar.github.io/hidden-states/
</code></pre>
</div>
<p>
BibTex:
</p>
<div>
<pre><code>@misc{alammar2021hiddenstates, 
  title={Finding the Words to Say: Hidden State Visualizations for Language Models},
  author={Alammar, J},
  year={2021},
  url={https://jalammar.github.io/hidden-states/}
}
</code></pre>
</div>
</div>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Jay Alammar – Visualizing machine learning one concept at a time. <br>
<br>
July 2, 2024 at 07:23PM
